src/fundamental/chid-fundamental.c

```diff
@@ -0,0 +1,130 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+
+/*
+ * Based on Nikita Travkin's dtbloader implementation.
+ * Copyright (c) 2024 Nikita Travkin <nikita@trvn.ru>
+ *
+ * https://github.com/TravMurav/dtbloader/blob/main/src/chid.c
+ */
+
+/*
+ * Based on Linaro dtbloader implementation.
+ * Copyright (c) 2019, Linaro. All rights reserved.
+ *
+ * https://github.com/aarch64-laptops/edk2/blob/dtbloader-app/EmbeddedPkg/Application/ConfigTableLoader/CHID.c
+ */
+
+#if SD_BOOT
+#  include "efi-string.h"
+#  define bswap_16(x) __builtin_bswap16(x)
+#  define bswap_32(x) __builtin_bswap32(x)
+#else
+#  include <string.h>
+#  include <uchar.h>
+#  include <byteswap.h>
+
+static inline size_t strlen16(const char16_t *s) {
+        if (!s)
+                return 0;
+
+        size_t len = 0;
+        while (*s) {
+                s++;
+                len++;
+        }
+
+        return len;
+}
+
+#endif
+
+#include "macro-fundamental.h"
+#include "chid-fundamental.h"
+#include "sha1-fundamental.h"
+
+static void get_chid(SmbiosInfo *info, uint32_t mask, Uuid *ret_chid) {
+        assert(mask);
+        assert(ret_chid);
+        uint8_t hash[SHA1_DIGEST_SIZE] = {};
+        const Uuid namespace = { 0x12d8ff70, 0x7f4c, 0x7d4c, {} }; /* Swapped to BE */
+
+        struct sha1_ctx ctx = {};
+        sha1_init_ctx(&ctx);
+
+        sha1_process_bytes(&namespace, &ctx, sizeof(namespace));
+
+        for (unsigned i = 0; i < CHID_SMBIOS_COUNT; i++)
+                if ((mask >> i) & 1) {
+                        if (i > 0)
+                                sha1_process_bytes(L"&", &ctx, 2);
+                        size_t len = strlen16(info->str[i]) * sizeof(*info->str[i]);
+                        if (len > 0)
```

Comment:
this conditoin seems unnecessary? sha1_process_bytes() sould treat len == 0 as a NOP on its own